### PR TITLE
[FIX] Fix bug in `fetch_single_file` with `resume=True` and improve tests

### LINKED ISSUE #4697: local_file never set to an actual value
I think this is a bug. These lines do not seem to be covered by the current tests, and `local_file` is set to `None` at Line ~574 and never changed after, so as far as I can tell the file opening should fail regardless of which function/method is used. I guess it is not that easy to test because it requires a partially downloaded file.

I am not familiar with what the function is supposed to do but if someone confirms that this is a bug then I can create an issue for it.

_Originally posted by @michellewang in https://github.com/nilearn/nilearn/pull/4696#discussion_r1815581252_

### PR DESCRIPTION
**Pull request by michellewang**
- Closes #4697

Changes proposed in this pull request:
- Use `request_mocker` so that the test actually works
- Deleted `local_file` (always `None`) and use actual filename instead
- Make exception catching more specific
- Some improvements in other tests, especially regarding to the use of context managers and `pathlib.Path`'s `read_text` and `write_text` methods

### CODE REVIEW

File: nilearn/datasets/tests/test_utils.py [hunk 1]
```diff
@@ -364,7 +364,7 @@ def test_uncompress_tar(tmp_path, ext, mode):
     # and check if ftemp exists
     ztemp = tmp_path / f"test.{ext}"
     ftemp = "test"
-    with contextlib.closing(tarfile.open(ztemp, mode)) as testtar:
+    with tarfile.open(ztemp, mode) as testtar:
```
Comment (michellewang) [inline nilearn/datasets/tests/test_utils.py:367]: Pretty sure the `contextlib.closing` is not needed : https://docs.python.org/3/library/tarfile.html#tarfile-objects

File: nilearn/datasets/tests/test_utils.py [hunk 2]
```diff
@@ -382,7 +382,7 @@ def test_uncompress_zip(tmp_path):
     # and check if ftemp exists
     ztemp = tmp_path / "test.zip"
     ftemp = "test"
-    with contextlib.closing(ZipFile(ztemp, "w")) as testzip:
+    with ZipFile(ztemp, "w") as testzip:
```
Comment (michellewang) [inline nilearn/datasets/tests/test_utils.py:385]: `Zipfile` is already a context manager: https://docs.python.org/3/library/zipfile.html#zipfile-objects

File: nilearn/datasets/tests/test_utils.py [hunk 3]
```diff
@@ -422,35 +422,57 @@ def test_safe_extract(tmp_path):
         _utils.uncompress_file(ztemp, verbose=0)
 
 
-def test_fetch_file_part(tmp_path):
+def test_fetch_single_file_part(tmp_path, capsys, request_mocker):
```
Comment (michellewang) [inline nilearn/datasets/tests/test_utils.py:425]: Looks like `fetch_single_file` used to be called `fetch_file`, in which case the tests should be renamed too

File: nilearn/datasets/_utils.py [hunk 4]
```diff
@@ -610,15 +609,15 @@ def fetch_single_file(
...
+                    with temp_full_name.open("ab") as fh:
                         _chunk_read_(
                             resp,
                             fh,
                             report_hook=(verbose > 0),
                             initial_size=initial_size,
                             verbose=verbose,
                         )
-            except Exception:
+            except OSError:
```
Comment (bthirion) [inline nilearn/datasets/_utils.py:620]: Can you test this line ?

**Review by bthirion — COMMENTED** [review]
Thx for the contribution.

File: nilearn/datasets/_utils.py (hunk 4 above)
Comment (michellewang) [reply nilearn/datasets/_utils.py:620]: Added a test for this

**Review by bthirion — APPROVED** [review]
LGTM, thx.

**Review by Remi-Gau — APPROVED** [review]

**Pull request merged**